### ISSUE TITLE
Provide 'duplicate' option against files in the sidebar

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#242]
Provide a file 'Duplicate' option through context-menu in sidebar. This suppose to receive two parameters:
- New name
- Target path (default is the same path)

**Comment by rat-moonshine** [comment]
After creating new/duplicated file I'm not opening it to the editor, but I'm open for suggestion. 

Known bug:
- Duplicate a MXML file to a folder outside of the project's source folder (i.e. inside bin-debug, or build directory)
- Try to open the newly created file after being created
- Language server reported problem in console:
`: Could not create compilation unit for file (final fallback): C:\Users\[user]\Desktop\NewFlexDesktopProject\obj\DuplicatedFile.mxml`
- Problem view shown following error:
`A fatal error occurred while checking for errors. Error checking disabled, except for simple syntax problems.`

@joshtynjala , do you have any immediate idea on this, or how it can be overcome (?)

**Comment by joshtynjala** [comment]
You are duplicating the file into a folder that isn't one of the project's source folders, so the language server is behaving as designed. 

I guess I could make the wire more clear. Instead of saying that it's simply a fatal error, I could modify it to say something like "MyFile.mxml is not located in the project's source path. Error checking disabled, except for simple syntax problems."

**Comment by joshtynjala** [comment]
Actually, I just reviewed my code in the language server and my explanation wasn't entirely accurate. The language server will actually try to provide code intelligence for files that are either in external source paths or *anywhere* in the project folder. That includes non-source folders in the project directory. Since this file is not in a source folder, the compiler that powers code intelligence doesn't know how to handle it.

I think I need to make the language server stricter so that folders in the project directory that aren't source paths are excluded.

**Comment by rat-moonshine** [comment]
I agree with the decision, @joshtynjala .

**Comment by rat-moonshine** [comment]
@joshtynjala , we're exploring options to update a duplicated ActionScript file name and its package details inside the file. 

Is there any way already exists to the language server which can update a class file accordingly, other than replacing any syntax that match the old file name?

**Comment by joshtynjala** [comment]
No, the language server doesn't have that capability.

**Comment by rat-moonshine** [comment]
Duplicating an ActionScript file now updates its class name and package path in following basic areas and conditions:

- If file duplicates to off-source folder, no action taken
- Checks and updates first occurrence of "package" reference
- Checks and updates first occurrence of "class [name]" 
- Checks and updates first occurrence of the constructor function

However, when the above updates may not be very sophisticated (in term of properly updating a file), this can work anyway now. If we see more basic test cases needs to cover, we can re-visit this later.

**Comment by JoelProminic** [comment]
I tested a variety of cases for duplicating a file or class:
- .mxml
- .xml
- .as to same directory
- .as to different package 
- .mxml to non-source directory
- .as to non-source directory

I did testing on both Flex Desktop and Royale projects.  The duplication and class update logic seems to be working well.

However, I noticed that if I entered the same name for the duplicate file, I got an error message for "Visual editor files" instead.

![image](https://user-images.githubusercontent.com/24532731/39780284-756206e6-52d1-11e8-9eea-b7f23128253c.png)
